File: tasks/TimeSeries-Downsampler/steps/milestone_2/tests/test_m2.py
"""Tests for Milestone 2: overlapping sliding windows."""
import os
import shutil
import subprocess
import tempfile

_MILESTONE = "m2"


def _run_rust_verifier():
    """Run the milestone's Rust assertions from a verifier-owned cargo harness.

    The harness is created in a temporary directory outside /app and depends on
    the agent's crate via a path dependency, so no verifier files are ever
    written into the agent's workspace. Returns the set of passing test names.
    """
    test_rs = os.path.join(
        os.path.dirname(os.path.abspath(__file__)), f"test_{_MILESTONE}.rs"
    )
    harness = tempfile.mkdtemp(prefix=f"verifier_{_MILESTONE}_")
    try:
        os.makedirs(os.path.join(harness, "src"), exist_ok=True)
        os.makedirs(os.path.join(harness, "tests"), exist_ok=True)
        with open(os.path.join(harness, "Cargo.toml"), "w") as fh:
            fh.write(
                "[package]\n"
                f'name = "verifier_{_MILESTONE}"\n'
                'version = "0.0.0"\n'
                'edition = "2021"\n\n'
                "[lib]\n"
                'path = "src/lib.rs"\n\n'
                "[dependencies]\n"
                'timeseries_downsampler = { path = "/app/app" }\n\n'
                "[[test]]\n"
                f'name = "test_{_MILESTONE}"\n'
                f'path = "tests/test_{_MILESTONE}.rs"\n'
            )
        open(os.path.join(harness, "src", "lib.rs"), "w").close()
        shutil.copyfile(
            test_rs, os.path.join(harness, "tests", f"test_{_MILESTONE}.rs")
        )
        result = subprocess.run(
            ["cargo", "test",
             "--manifest-path", os.path.join(harness, "Cargo.toml"),
             "--test", f"test_{_MILESTONE}", "--offline",
             "--", "--test-threads=1"],
            capture_output=True, text=True, timeout=300,
        )
        output = result.stdout + result.stderr
        passed = set()
        for line in output.splitlines():
            line = line.strip()
            parts = line.split(" ... ")
            if line.startswith("test ") and len(parts) == 2 and parts[1].strip() == "ok":
                passed.add(parts[0][5:])
        return passed
    except Exception:
        return set()
    finally:
        shutil.rmtree(harness, ignore_errors=True)


_PASSED = _run_rust_verifier()

class TestMilestone2:
    def test_single_event_populates_multiple_windows(self):
        """A single event is added to all overlapping valid sliding windows."""
        assert "single_event_populates_multiple_windows" in _PASSED

    def test_sliding_window_boundaries_correct(self):
        """Events strictly inside the valid boundaries are included, while edges are correctly handled (start inclusive, end exclusive)."""
        assert "sliding_window_boundaries_correct" in _PASSED

    def test_multiple_events_aggregate_across_slides(self):
        """Multiple events aggregate correctly when their valid windows overlap."""
        assert "multiple_events_aggregate_across_slides" in _PASSED

    def test_duplicate_events_apply_bitwise_xor(self):
        """Duplicate events in the same window are aggregated using bitwise XOR of their IEEE 754 representations."""
        assert "duplicate_events_apply_bitwise_xor" in _PASSED

    def test_median_computation_even(self):
        """Median computation correctly averages the two middle elements when N is even."""
        assert "median_computation_even" in _PASSED

File: tasks/TimeSeries-Downsampler/steps/milestone_2/tests/test_m2.rs
use timeseries_downsampler::models::Event;
use timeseries_downsampler::sliding::SlidingWindowAggregator;

#[test]
fn single_event_populates_multiple_windows() {
    let mut agg = SlidingWindowAggregator::new(300000, 60000);
    let event = Event { timestamp_ms: 300000, event_id: "e1".to_string(), name: "mem".to_string(), value: 10.0, dependency_ids: Vec::new() };
    agg.add_event(&event);
    
    assert_eq!(agg.flush_window(0).len(), 0);
    
    let w60 = agg.flush_window(60000);
    assert_eq!(w60.len(), 1);
    assert_eq!(w60[0].window_start_ms, 60000);
    assert_eq!(w60[0].window_end_ms, 360000);
    assert_eq!(w60[0].name, "mem");
    assert_eq!(w60[0].avg, 10.0);
    assert_eq!(w60[0].count, 1);

    assert_eq!(agg.flush_window(120000).len(), 1);
    assert_eq!(agg.flush_window(180000).len(), 1);
    assert_eq!(agg.flush_window(240000).len(), 1);
    assert_eq!(agg.flush_window(300000).len(), 1); 
    assert_eq!(agg.flush_window(300000).len(), 0); // verify cleared
    assert_eq!(agg.flush_window(360000).len(), 0); 
}

#[test]
fn sliding_window_boundaries_correct() {
    let mut agg = SlidingWindowAggregator::new(100, 50);
    let event = Event { timestamp_ms: 100, event_id: "e1".to_string(), name: "mem".to_string(), value: 10.0, dependency_ids: Vec::new() };
    agg.add_event(&event);
    
    assert_eq!(agg.flush_window(0).len(), 0);
    assert_eq!(agg.flush_window(50).len(), 1);
    assert_eq!(agg.flush_window(100).len(), 1);
}

#[test]
fn multiple_events_aggregate_across_slides() {
    let mut agg = SlidingWindowAggregator::new(100, 50);
    agg.add_event(&Event { timestamp_ms: 60, event_id: "e1".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: Vec::new() });
    agg.add_event(&Event { timestamp_ms: 90, event_id: "e2".to_string(), name: "cpu".to_string(), value: 20.0, dependency_ids: Vec::new() });
    
    let r0 = agg.flush_window(0);
    assert_eq!(r0[0].count, 2);
    assert_eq!(r0[0].avg, 15.0);
    assert!((r0[0].median - 15.0).abs() < 1e-9); // N=2, Median of 10.0 and 20.0 is 15.0
    assert_eq!(r0[0].min, 10.0);
    assert_eq!(r0[0].max, 20.0);
    
    let r50 = agg.flush_window(50);
    assert_eq!(r50[0].count, 2);
    assert_eq!(r50[0].avg, 15.0);
    
    let r100 = agg.flush_window(100);
    assert_eq!(r100.len(), 0);
}

#[test]
fn duplicate_events_apply_bitwise_xor() {
    let mut agg = SlidingWindowAggregator::new(100, 50);
    agg.add_event(&Event { timestamp_ms: 60, event_id: "e1".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: Vec::new() });
    agg.add_event(&Event { timestamp_ms: 70, event_id: "e1".to_string(), name: "cpu".to_string(), value: 15.0, dependency_ids: Vec::new() }); 
    
    // 10.0 in f64 bits XOR 15.0 in f64 bits
    let v1 = 10.0f64.to_bits();
    let v2 = 15.0f64.to_bits();
    let expected_xor = f64::from_bits(v1 ^ v2);
    
    agg.add_event(&Event { timestamp_ms: 90, event_id: "e2".to_string(), name: "cpu".to_string(), value: 15.0, dependency_ids: Vec::new() });
    
    let r0 = agg.flush_window(0);
    assert_eq!(r0[0].count, 2); // e1 and e2
    assert_eq!(r0[0].avg, (expected_xor + 15.0) / 2.0);
}

#[test]
fn median_computation_even() {
    let mut agg = SlidingWindowAggregator::new(100, 50);
    for i in 1..=20 {
        agg.add_event(&Event { timestamp_ms: 60, event_id: format!("e{}",i), name: "cpu".into(), value: i as f64, dependency_ids: Vec::new() });
    }
    let r = agg.flush_window(0);
    // Median of 1..=20 is (10 + 11) / 2 = 10.5
    assert!((r[0].median - 10.5).abs() < 1e-9);
}
